game/scripts/vscripts/components/surrender/surrender.lua

```diff
@@ -0,0 +1,147 @@
+  end
+end
+
+function SurrenderManager:ScoreAllowsSurrender(teamId)
+  local score = CustomNetTables:GetTableValue('team_scores', 'score')
+  local scoreDiff
+  if teamId == DOTA_TEAM_GOODGUYS then
+    scoreDiff = score.badguys - score.goodguys
+  elseif teamId == DOTA_TEAM_BADGUYS then
+    scoreDiff = score.goodguys - score.badguys
+  end
+
+  return scoreDiff >= MINIMUM_KILLS_BEHIND
+end
+
+function SurrenderManager:TimeAllowsSurrender(playerId, now)
+  local result
+  if lastTimeSurrenderWasCalled == nil then
+    DebugPrint("SurrenderManager:TimeAllowsSurrender --> lastTimeSurrenderWasCalled == nil")
+    result = true
+  else
+    if now - lastTimeSurrenderWasCalled > ONE_MINUTE then
+      DebugPrint("SurrenderManager:TimeAllowsSurrender --> now - lastTimeSurrenderWasCalled = " .. (now - lastTimeSurrenderWasCalled))
+      if lastTimeSurrenderWasCalledByPlayer[playerId] == nil or
+          now - lastTimeSurrenderWasCalledByPlayer[playerId] > FIVE_MINUTES then
+        if lastTimeSurrenderWasCalledByPlayer[playerId] == nil then
+          DebugPrint("SurrenderManager:TimeAllowsSurrender --> lastTimeSurrenderWasCalledByPlayer[playerId] == nil")
+        else
+          DebugPrint("SurrenderManager:TimeAllowsSurrender --> now - lastTimeSurrenderWasCalled = " .. (now - lastTimeSurrenderWasCalledByPlayer[playerId]))
+        end
+        result = true
+      else
+        result = false
+      end
+    else
+      result = false
+    end
+  end
+
+  return result
+end
+
+function SurrenderManager:PlayerVote (eventSourceIndex, args)
+  local playerId = eventSourceIndex.PlayerID
+  local selection = eventSourceIndex.result
+  votes[playerId] = selection
+  numberOfVotesCast = numberOfVotesCast + 1
+  DebugPrint("numberOfVotesCast = " .. numberOfVotesCast)
+  if numberOfVotesCast == numberOfVotesExpected then
+    SurrenderManager:CalculateVotes()
+  end
+end
+
+function SurrenderManager:CalculateVotesFromTimer()
+DebugPrint("numberOfVotesExpected = " .. numberOfVotesExpected)
+  if numberOfVotesExpected > 0 then -- if CalculateVotes has already ran this will be 0
+    SurrenderManager:CalculateVotes()
+  end
+end
+
+function SurrenderManager:CalculateVotes()
+  local yesVotesCast = 0
+  for key,value in pairs(votes) do
+    yesVotesCast = yesVotesCast + value
+  end
+
+  DebugPrint("yesVotesCast = " .. yesVotesCast)
+  if table.getn{REQUIRED_YES_VOTES} <= numberOfVotesCast then
+    local requiredNumberOfYesVotes = REQUIRED_YES_VOTES[numberOfVotesCast]
+    votes = {}
+    numberOfVotesCast = 0
+    numberOfVotesExpected = 0
+    DebugPrint("requiredNumberOfYesVotes = " .. requiredNumberOfYesVotes)
+    if requiredNumberOfYesVotes <= yesVotesCast then
+      DebugPrint("End game")
+      local teamText = nil
+      if teamIdToSurrender == DOTA_TEAM_GOODGUYS then
+        teamText = "Radiant"
+      elseif teamIdToSurrender == DOTA_TEAM_BADGUYS then
+        teamText = "Dire"
```

Comment:
I feel like the team names shouldn't be hardcoded. Unsure what the best way to handle this is. One way would be to just have completely separate keys. Not sure if the system localises the DialogVariable stuff as well. Could try I suppose. Vanilla Dota already has keys for "Radiant" and "Dire", "DOTA_GoodGuysShort" and "DOTA_BadGuysShort", respectively.

Reply:
I cannot find these constants - I'm thinking they are unavailable in the mod(?)
Should I add these constants? Where do they go?

Reply:
They're in vanilla Dota, they can be used without any need to put them in our localisation files. https://raw.githubusercontent.com/SteamDatabase/GameTracking-Dota2/master/game/dota/resource/dota_english.txt

Reply:
I tried using them and it did not work, do I need to declare something (usage\import etc.)?

Reply:
Anyone?

Reply:
No. If you put them in the variable part and it didn't work then it just means those don't get run through localisation, which is unfortunate. The other option is just to have keys for surrender text for both teams. Similar to the victory messages in vanilla Dota.

Reply:
Is it possible to merge as is and I'll change the text later? - I tried to change it, then run the translation but it had some errors. Maybe I need to rebase, but I'm new to the git environment and not sure I'll have time to research how to do that here - I'm used to having the option in the UI and can't seem to find it in the GitHub Desktop.